relay-sampling/src/evaluation.rs

```diff
@@ -65,6 +65,11 @@ impl<'a> ReservoirEvaluator<'a> {
         }
     }
 
+    /// Gets shared ownership of the reference counters.
```

Comment:
```suggestion
    /// Gets shared ownership of the reservoir counters.
```

Is there any invariants to uphold if someone gets access to the counters (if so please add them to the docs)?

E.g. counters need to only be update when X and then you also need to keep Y in sync.

Reply:
I see now you're just using it clone the envelope into a new one, maybe we should have a constructor which allows you to "fork off" an envelope from an existing one instead instead of exposing all the internals one by one.

Reply:
> maybe we should have a constructor which allows you to "fork off" an envelope

I tried this (as a method on `ProcessEnvelopeState`) but it did not really improve semantics, because the fork requires some properties from the envelope, some from the state, and some from the processor itself. Maybe this will become more straight-forward when we clean up / merge `ManagedEnvelope` with `TypedEnvelope` and `ProcessEnvelopeState`.